### ISSUE TITLE
Job cattle-node-cleanup fails due to restrictive standard PSP

### ISSUE THREAD

**Issue by pwurbs** [rancher/rancher#36006]
**Rancher Server Setup**
- Rancher version: 2.6.3

**Information about the Cluster**
- Kubernetes version: v1.21.7
- Cluster Type: Downstream
- Custom cluster / Provider: RKE1

**Describe the bug**
There is the periodic event in the cluster event log:
`FailedCreate 	Job cattle-node-cleanup-xxx
Error creating: pods "cattle-node-cleanup-xxx-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used] `

**To Reproduce**
- Remove a node from a downstream cluster
- Watch the job cattle-node-cleanup-xxx
--> Recent events: 
Error creating: pods "cattle-node-cleanup-xxx" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used] 

**Result**
Error creating: pods "cattle-node-cleanup-xxx" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used] 

**Expected Result**
- Job cattle-node-cleanup-xxx should run without errors
- The job should not run in the default namespace, because it violates the CIS policy

**Additional context**
I use a more restrictive standard PSP in the cluster which prohibits the usage of a hostPath volume.
The YAML file of the job "cattle-node-cleanup-8nx54" shows, that no specific service account is used. So the default service account is used which is bound to my default PSP for the cluster.
The job should use a specific service account to which any unrestricted PSP is bound.

SURE-3708

**Comment by pwurbs** [comment]
Addition: And the job should not be performed in the Default namespace, because this is against the CIS scan rules.

**Comment by pwurbs** [comment]
Workaround: Remove the failing job. Then the periodic event in the cluster log vanishes. If the node is removed anyway, it possibly doesn't matter, if the cleaning job is not performed.

**Comment by Martin-Weiss** [comment]
> Workaround: Remove the failing job. Then the periodic event in the cluster log vanishes. If the node is removed anyway, it possibly doesn't matter, if the cleaning job is not performed.

I am not sure if this is a valid workaround as the cleanup might not only delete the node object but also other things?

**Comment by pwurbs** [comment]
> > Workaround: Remove the failing job. Then the periodic event in the cluster log vanishes. If the node is removed anyway, it possibly doesn't matter, if the cleaning job is not performed.
> 
> I am not sure if this is a valid workaround as the cleanup might not only delete the node object but also other things?

Yes, but if the underlying server (at cloud level) is removed anyway, then it should not matter. But of course not quite sure...

**Comment by Martin-Weiss** [comment]
Do we know if the cleanup only deletes stuff on the "local to the deleted server"? Probably there are also some other k8s objects that might need to be deleted?

I believe for the workaround we should find out which serviceaccount is used in the job and then create a role/rolebinding that would allow usage of system-unrestricted-psp for the used serviceaccounts.

I.e.

```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: cattle-cleanup-role
  namespace: default
rules:
- apiGroups:
  - policy
  resourceNames:
  - system-unrestricted-psp
  resources:
  - podsecuritypolicies
  verbs:
  - use
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: cattle-cleanup-rolebinding
  namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: cattle-cleanup-role
subjects:
- kind: ServiceAccount
  name: cattle-cleanup-c22lc
  namespace: default
- kind: ServiceAccount
  name: cattle-cleanup-f5v84
  namespace: default
```

**Comment by pwurbs** [comment]
> Do we know if the cleanup only deletes stuff on the "local to the deleted server"? Probably there are also some other k8s objects that might need to be deleted?

Unfortunately not

> I believe for the workaround we should find out which serviceaccount is used in the job and then create a role/rolebinding that would allow usage of system-unrestricted-psp for the used serviceaccounts.

The job currently uses the default service account. Therefore my strict PSP applies. And it's deployed in the default namespace. 
And yes, the creation of appropriate SA, Role, Rolebinding would be the right way forward. But I wouldn't know where the manifests are stored or which instance triggers the creation of the job. 
And the namespace should changed to cattle-system

**Comment by Martin-Weiss** [comment]
I have no idea how to fix this in general (I am not a developer  ;-)) - but the job should run in case the role/roleassignment would be done with the above yaml (adjusted to default service account) and then delete the role/rolebinding after the job completed.

**Comment by pwurbs** [comment]
Only Rancher can and should fix this. To be honest, it's a pitty, that the vendor doesn't consider such basic things like service accounts and namespaces for such cleaning jobs.
But thanks for your proposal. It seems to be a cumbersome but valid workaround. I will document it and apply it the next time...

**Comment by Martin-Weiss** [comment]
> Only Rancher can and should fix this. To be honest, it's a pitty, that the vendor doesn't consider such basic things like service accounts and namespaces for such cleaning jobs. But thanks for your proposal. It seems to be a cumbersome but valid workaround. I will document it and apply it the next time...

True - a real fix should be implemented in Rancher. 
Could you create a Support Case so that this gets tracked properly?

**Comment by pwurbs** [comment]
Sorry, I am not under maintenance. Therefore I created this Github issue...

**Comment by pwurbs** [comment]
This issue is still relevant and not solved at all. Please leave it open...

**Comment by sirredbeard** [comment]
@rosskirkpat has requested some additional information from the customer via the internal support issue.

@luthermonson states `All our cleanup jobs across all of rancher (cluster/node) are in the default namespace as it's the only thing guaranteed to be there. This was a design decision made quite some time ago.`

**Comment by pwurbs** [comment]
The default namespace is not the problem. The problem is, that the job doesn't start due to my restricted PSP. The creation and assignment of a SA with appropriate permissions should be sufficient.

**Comment by deniseschannon** [comment]
With the recent team changes, I'd like this re-triaged and re-prioritized in the team to see when and who should pick it up

**Comment by paynejacob** [comment]
Moving to team 1 as the root cause is related to the PSP assigned to the cleanup job.

**Comment by HoustonDad** [comment]
Can we keep this targeted for 2.6? We have a lot of clients hitting this and it will be a while before they can update to 2.7.x

Thanks!

**Comment by cbron** [comment]
This doesn't appear to deal at all with the Rancher object/chart itself, or its PSPs. This is the `cattle-node-cleanup` pod, which needs a service-account+PSP attached to it in order to function in a restricted environment: https://github.com/rancher/rancher/blob/release/v2.7/pkg/controllers/management/node/cleanup.go#L279-L280

Or as mentioned above it could run in the system namespace, which is probably more appropriate. I don't understand why the cattle system namespaces wouldn't exist, even in rke1 standalone ?

**Comment by annablender** [comment]
@pwurbs I am working on reproducing this issue. Could you please share the yaml file for your restricted standard PSP? Or at the structure with redacted data? Thank you

**Comment by pwurbs** [comment]
@annablender 
Here it is:
```
spec:
  allowPrivilegeEscalation: false
  defaultAddCapabilities:
    - CHOWN
  defaultAllowPrivilegeEscalation: false
  fsGroup:
    rule: RunAsAny
  runAsGroup:
    rule: RunAsAny
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  volumes:
    - emptyDir
    - gitRepo
    - secret
    - persistentVolumeClaim
    - downwardAPI
    - configMap
    - projected
```

**Comment by annablender** [comment]
@pwurbs I was unable to reproduce this issue on 2.6.3.

Repro steps
* Run rancher instance v2.6.3
* Create rke1 downstream cluster 1 cp,etcd 2 worker nodes with PSP enabled and set to `restricted-test` using the config file above
* Remove a downstream worker node
* Node is removed immediately. Cattle-node-cleanup-xxx does not get stuck and the node is removed too fast for me to see any pod logs

<img width="529" alt="image" src="https://user-images.githubusercontent.com/12955547/219091632-543cc04a-4433-4fd6-8312-e3205ff48154.png">

<img width="647" alt="image" src="https://user-images.githubusercontent.com/12955547/219091118-e99de70e-2e1d-42d3-b013-aacd396bbf17.png">

Are you still seeing this issue?

**Comment by pwurbs** [comment]
@annablender You are right. The issue must have been solved by one of the latest updates. In 2.6.10 (and maybe former versions) I am not facing this issue. 
But apart from that, would you like to check, in which namespace the cleanup pod runs now? It was my main concern that it runs in default namespace, which was also against the CIS rules.

**Comment by annablender** [comment]
@pwurbs According to @luthermonson https://github.com/rancher/rancher/issues/36006#issuecomment-1095244459 all cleanup jobs including `cattle-node-cleanup-xxx` run in the default namespace. They are guaranteed to be there. I know there's a Rancher 2.6 and 2.7 [hardening guide](https://ranchermanager.docs.rancher.com/pages-for-subheaders/cis-scans) for CIS but could you link me the docs stating jobs/cleanup jobs should run in a non-default namespace?

**Comment by pwurbs** [comment]
@annablender Here is the link: https://ranchermanager.docs.rancher.com/v2.6/reference-guides/rancher-security/rancher-v2.6-hardening-guides/rke1-self-assessment-guide-with-cis-v1.6-benchmark#574-the-default-namespace-should-not-be-used-automated

**Comment by annablender** [comment]
@pwurbs Thank you! I will discuss this with my team. Otherwise, QA can test this using repro steps https://github.com/rancher/rancher/issues/36006#issuecomment-1431649029.

**Comment by andyblendermann** [comment]
@pwurbs Thanks for the link. From the discussion, the `cattle-node-cleanup-xxx` jobs run in the default ns because that is a ns that is most accessible to rancher, and will not be removed during the cleanup process. Rancher cleans up the `cattle-system` ns and others on custom clusters and it would not complete the process successfully if rancher removed the ns the cleanup job was in.

I also will try again to reproduce this on a _custom cluster_, not a node provisioned one since the cleanup jobs only run on the former. That's my mistake. This is still in review.

**Comment by a-blender** [comment]
@sowmyav27 This is ready to test.

I further investigated this issue and was able to reproduce it on rancher 2.6.2 but was unable to reproduce it on rancher `2.6.10.` As per this comment https://github.com/rancher/rancher/issues/36006#issuecomment-1434247590, I believe this was fixed somewhere in between those versions.

Please have QA verify this is fixed on the latest rancher release `v2.6.11.`

## Testing template

#### Root cause
 
#### What was fixed, or what changes have occurred

#### Areas or cases that should be tested

Test steps
* Run rancher instance latest v2.6.11
* Create an rke1 custom cluster with PSP restricted enabled, 3 nodes (1 cp/etcd, 2 worker)
* Cordon, drain, and then delete a worker node from k8s

What you are looking for

1) cattle-node-cleanup-xxx job to be _successfully_ deployed on the node without a `HostPath` error in the job yaml

```
kubectl get jobs --kubeconfig
NAME                        COMPLETIONS   DURATION   AGE
cattle-node-cleanup-mjqf2   1/1                      15h
cattle-node-cleanup-ztfk6   1/1                       17m
```

2) node removed successfully from the rancher UI

* Repeat with second worker node. Verify same behavior.
 
#### What areas could experience regressions ?

RKE1 prov
 
#### Are the repro steps accurate/minimal ?

Yes.

**Comment by vivek-shilimkar** [comment]
Validated the issue on rancher v2.7.5 with k8s version v1.23 with PSP enabled and set to restricted. Issue is not active anymore.

Validations steps.
Test steps

1. Created a rancher v2.7.5 (this version still has PSP enabled.)
2. Created an rke1 custom cluster with PSP enabled and set to restricted, 3 nodes (1 cp/etcd, 2 worker)
3. Cordon, drain, and then delete a worker node from rancher.
4. cattle-node-cleanup-xxx job was deployed successfully on the cluster.
5. Node deletion was successful.

Issue was not encountered, hence closing the issue.
